File: LCA.js
const push = (node, val) => {
  if (!node) return { val: val, left: undefined, right: undefined };
  else if (val < node.val) node.left = push(node.left, val);
  else if (val > node.val) node.right = push(node.right, val);
  return node;
};

const LCA = (root, v, w) => {
  const path1 = [];
  const path2 = [];
  if (!findPath(root, v, path1) || !findPath(root, w, path2)) return false;
  for (let i = 0; i < path1.length || i < path2.length; i++)
    if (path1[i] != path2[i]) return path1[i - 1];
};

const findPath = (root, n, path) => {
  if (!root) return false;
  path.push(root.val);
  if (root.val === n) return true;
  if (root.left && findPath(root.left, n, path)) return true;
  if (root.right && findPath(root.right, n, path)) return true;
  path.pop();
  return false;
};

let root = undefined;
for (let i = 0; i < 8; i++) {
  let val = (Math.random() * 8) | 0;
  console.log(val);
  root = push(root, val);
}

const print = (root, n) =>
  `(${root.left ? print(root.left, ++n) : ""}) + ${
    root.val + (n === 0 ? "r" : "")
  } + (${root.right ? print(root.right, ++n) : ""})`;
console.log(print(root, 0));
console.log(LCA(root, 4, 5));

File: LCA.go
package main

import "fmt"

type node struct {
	left  *node
	right *node
	val   int
}

func push(val int, array []int, index *int) {
	array[*index] = val
	*index++
}

func pop(index *int) {
	*index--
}

func add(root *node, val int) *node {

	if root == nil {
		return &node{nil, nil, val}
	}
	if val < root.val {
		root.left = add(root.left, val)
	} else if val > root.val {
		root.right = add(root.right, val)
	}
	return root
}

func findPath(root *node, n int, path []int, index *int) bool {
	if root == nil {
		return false
	}
	push(root.val, path, index)
	if root.val == n {
		return true
	}
	if root.left != nil && findPath(root.left, n, path, index) {
		return true
	}
	if root.right != nil && findPath(root.right, n, path, index) {
		return true
	}
	pop(index) 
	return false
}

func LCA(root *node, v int, w int) int {
	var arr0 [100]int
	var arr1 [100]int

	var index0 int = 0
	var index1 int = 0

	if !findPath(root, v, arr0[0:100], &index0) || !findPath(root, w, arr1[0:100], &index1) {
		return -1
	}
	fmt.Println(arr0[0:index0])
	fmt.Println(arr1[0:index1])
	for i := 0; i < index0 && i < index1; i++ {
		if arr0[i] != arr1[i] {
			return arr0[i-1]
		}
	}
	return -1
}
func _print(root *node) string {
  left := ""
  if root.left != nil {
    left = _print(root.left)
  }
  right := ""
  if root.right != nil {
    right = _print(root.right)
  }
  return fmt.Sprint("(", left, ")", root.val, "(", right, ")")
}

func doSomething(array []int){
  for i := 0; i < 10; i++ {
    array[i] = 42
  }
}
func main() {
	var root *node = nil
	for i := 0; i < 8; i++ {
		root = add(root, i)
	}
	root = add(root, -2)
  fmt.Println(_print(root))
	fmt.Println(LCA(root, 1, -2))
}
